src/lib/plugin.js

```diff
@@ -134,37 +125,62 @@ module.exports = class PluginVirtual extends EventEmitter2 {
     }
   }
 
-  * _connect () {
+  registerRequestHandler (handler) {
+    if (this._requestHandler) {
+      throw new RequestHandlerAlreadyRegisteredError('requestHandler is already registered')
+    }
+
+    if (typeof handler !== 'function') {
+      throw new InvalidFieldsError('requestHandler must be a function')
+    }
+
+    this._requestHandler = handler
+  }
+
+  deregisterRequestHandler () {
+    this._requestHandler = null
+  }
+
+  async connect () {
     // read in from the store and write the balance
-    yield this._balance.connect()
+    await this._balance.connect()
 
     this._connected = true
     this._safeEmit('connect')
   }
 
-  * _disconnect () {
+  async disconnect () {
     this._connected = false
     this._safeEmit('disconnect')
   }
 
-  * _sendMessage (message) {
+  async sendRequest (message) {
     this._validator.validateOutgoingMessage(message)
-    yield this._rpc.call('send_message', this._prefix, [message])
+    this._safeEmit('outgoing_request', message)
+
+    const response = await this._rpc.call('send_request', this._prefix, [message])
+    this._validator.validateIncomingMessage(response)
+    this._safeEmit('incoming_response', response)
 
-    this._safeEmit('outgoing_message', message)
+    return response
   }
 
-  * _handleMessage (message) {
+  async _handleRequest (message) {
     this._validator.validateIncomingMessage(message)
+    this._safeEmit('incoming_request', message)
 
-    // assign legacy account field
-    this._safeEmit('incoming_message', Object.assign({},
-      message,
-      { account: this._prefix + this._peerPublicKey }))
-    return true
+    if (!this._requestHandler) {
+      throw new NotAcceptedError('no request handler registered')
+    }
+
+    const response = await this._requestHandler(message)
```

Comment:
I think `_requestHandler` needs a `.catch()` that returns an error response message (see https://github.com/interledgerjs/ilp-plugin-bells/blob/fb11472942a220537edf49be0911a7a4f227bfe8/src/lib/plugin.js#L428-L439).